go/engine/pgp_keygen.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
+
+import (
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+)
+
+// PGPKeyGen is an engine.
+type PGPKeyGen struct {
+	libkb.Contextified
+	arg keybase1.PGPKeyGenDefaultArg
+}
+
+// NewPGPKeyGen creates a PGPKeyGen engine.
+func NewPGPKeyGen(g *libkb.GlobalContext, arg keybase1.PGPKeyGenDefaultArg) *PGPKeyGen {
+	return &PGPKeyGen{
+		Contextified: libkb.NewContextified(g),
+		arg:          arg,
+	}
+}
+
+// Name is the unique engine name.
+func (e *PGPKeyGen) Name() string {
+	return "PGPKeyGen"
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *PGPKeyGen) Prereqs() Prereqs {
+	return Prereqs{
+		Device: true,
+	}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *PGPKeyGen) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.PgpUIKind,
+		libkb.SecretUIKind,
+	}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *PGPKeyGen) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{
+		&PGPKeyImportEngine{},
+	}
+}
+
+// Run starts the engine.
+func (e *PGPKeyGen) Run(ctx *Context) error {
+
+	// generate a new pgp key with defaults (and no push)
+	arg := PGPKeyImportEngineArg{
+		Ctx:        e.G(),
+		AllowMulti: true,
+		OnlySave:   true,
+		Gen:        &libkb.PGPGenArg{Ids: libkb.ImportPGPIdentities(e.arg.CreateUids.Ids)},
+	}
+	eng := NewPGPKeyImportEngine(arg)
+	if err := RunEngine(eng, ctx); err != nil {
+		return err
+	}
+
+	// tell the UI about the key
+	e.G().Log.Debug("generated pgp key: %s", eng.bundle.GetFingerprint())
+	pub, err := eng.bundle.Encode()
+	if err != nil {
+		return err
+	}
+	keyArg := keybase1.KeyGeneratedArg{
+		Kid: eng.bundle.GetKID(),
+		Key: keybase1.KeyInfo{
+			Fingerprint: eng.bundle.GetFingerprint().String(),
+			Key:         pub,
+			Desc:        eng.bundle.VerboseDescription(),
+		},
+	}
+	if err := ctx.PgpUI.KeyGenerated(ctx.NetContext, keyArg); err != nil {
+		return err
+	}
+
+	// ask if we should push it to api server
+	push, err := ctx.PgpUI.ShouldPushPrivate(ctx.NetContext, ctx.SessionID)
```

Comment:
Well, we might advise the front end not to even support this feature, since we don't want to be asking passwords all of the sudden again....

If this is a feature that's going to be prominent in the UI, then we should consider the other syncing system, that uses encryption device keys rather than a stretched password.